Align SearchFilter behaviour to django.contrib.admin search

sevdog opened the pull request:
## Description
Make `SearchFilter` behave more like `django.contrib.admin` search (which is already stated in docs). 

- Allow use any lookup by appending it the fields in `search_fields`
- Perform a query-only distinct filtering
- Handle quotes surrounded terms with escapes (ie: `"the text"`, `"the \"text\""`)
- Split by spaces only by spaces, considering quoteing (**breaking change**)
- Raise a validation error if any null-character is provided in search (**breaking change**)

See #8205

auvipy commented:
looks like a nice improvment

auvipy commented on tests/test_filters.py:
```diff
@@ -177,6 +182,7 @@ class SearchListView(generics.ListAPIView):
 
         request = factory.get('/', {'search': r'^\w{3}$', 'title_only': 'true'})
         response = view(request)
+        print(response.data)
```
```suggestion
```

auvipy replied:
I guess we could remove it as it seems to be added on debug purpose only?

sevdog replied:
Yeah, sorry.

auvipy commented:
can we avoid at least one breaking change 

- Split by spaces only by spaces, considering quoteing (breaking change) --- and keep both? if it is not too hard?

sevdog commented:
> can we avoid at least one breaking change
> 
> * Split by spaces only by spaces, considering quoteing (breaking change) --- and keep both? if it is not too hard?

This could be interesting, however it should be defined the order of split logics due to how `smart_split` works and which kind of user experience is expected. Just to give an example, consider the following search:

```python
'"hello, world" can I,help you'
```

With _only_ `smart_split` this becomes 
```python
['hello, world', 'can', 'I,help', 'you']
```

Maybe a good compromise would be:
- if comma is inside quotes, do not split
- otherwhise split

With this logic the first search whould become

```python
['hello, world', 'can', 'I' , 'help', 'you']
```

There is an _edge case_: what to do if there are one or more commas before or after a quoted sentence?

```python
',"hello world", what to,,do,?'
```

I belive that in this case it would be nice to _strip_ commas from tokens extracted by `smart_split` and then yield quoted sentences as the are and split other by comma.

The generator to handle this iteration whould be something like:

```python
def search_smart_split(search_terms):
    for term in smart_split(search_terms):
        term = term.strip(',')
        if term.startswith(('"', "'")) and term[0] == term[-1]:
            yield unescape_string_literal(term)
        else:
            yield from (sub_term for sub_term in term.split(',') if sub_term)
```

This should parse the latter term into:
```python
['hello world', 'what', 'to', 'do', '?']
```

sevdog commented:
Testing the previous approach I encountered an other case which I did not considered before. What to do when there is a quoted phrase followed by other characters:

```python
'"hello, world",found'
```

The `smart_split` keeps this phrase togheter, so splitting by comma results in

```python
['"hello', ' world"', 'found']
```

The whitespace is easy and intuitive to remove (IMO), but I am not sure if in this case also the quotes should be removed (thus using `.split(' \'"')` instead of `.split()`).

auvipy commented:
are we going to preserve the old way or going to have option for both new and old ways?

sevdog commented:
> are we going to preserve the old way or going to have option for both new and old ways?

It depends on which mechanism are we willing to provide to select this behaviour. I can think these:

- an attribute on filter class
- an attribute on view
- a settings (which would be global for every view/filter)

auvipy commented:
from the tests it seems that both type of search is supported in this PR, right? I would like to consider this improvement for 3.15 release

sevdog commented:
> from the tests it seems that both type of search is supported in this PR, right? I would like to consider this improvement for 3.15 release

Yes, `search_smart_split` does first the split like django.contrib.admin and then splits by comma. The only case in which this would give different results is where the search input is a mixed combination of quoted strings and commas. A simple CSV vaue (ie: `an,example,search`) would be splitted like it was before.

auvipy commented on rest_framework/compat.py:
```diff
@@ -14,13 +13,6 @@ def unicode_http_header(value):
     return value
 
 
-def distinct(queryset, base):
```
can we change this in a separate PR? or it is part of the change in this PR?

sevdog replied:
It could be splitted, however this method was introduced in https://github.com/encode/django-rest-framework/commit/b4b2dc18fa4cb6bbe008d4f6e59b58862ed20be6 only for the search class and putted into `compact` package in the hope of being re-used.

IMO this should be kept togheter with the rest of the PR since this change goes in the same direction of "align the behaviour with django".

I just noticed that the code I used (which was introduced in https://github.com/django/django/commit/187118203197801c6cb72dc8b06b714b23b6dd3d) has just been removed from the main branch in https://github.com/django/django/commit/d569c1dcfeb26ca9ee391e5dfeadedf2b5ed4253. So the whole thing may just be changed to use `.distinct()` like django is going to do.

PS: I really do not understand why 8 years ago there was the need to have a different implementation of distinct only for Oracle engine and why this was implemented here.

auvipy replied:
> PS: I really do not understand why 8 years ago there was the need to have a different implementation of distinct only for Oracle engine and why this was implemented here.

at that time django oracle back end was not properly maintained. but now it is

auvipy commented on docs/api-guide/filtering.md:
```diff
@@ -213,12 +213,12 @@ This will allow the client to filter the items in the list by making queries suc
     search_fields = ['data__breed', 'data__owner__other_pets__0__name']
 
-By default, searches will use case-insensitive partial matches.  The search parameter may contain multiple search terms, which should be whitespace and/or comma separated.  If multiple search terms are used then objects will be returned in the list only if all the provided terms are matched.
+By default, searches will use case-insensitive partial matches.  The search parameter may contain multiple search terms, which should be whitespace separated.  If multiple search terms are used then objects will be returned in the list only if all the provided terms are matched.
```
do we need further enhancement of the docs?

sevdog replied:
I will add some examples about search term and how it will get parsed.

auvipy commented on docs/api-guide/filtering.md:
```diff
@@ -213,19 +213,23 @@ This will allow the client to filter the items in the list by making queries suc
     search_fields = ['data__breed', 'data__owner__other_pets__0__name']
 
-By default, searches will use case-insensitive partial matches.  The search parameter may contain multiple search terms, which should be whitespace and/or comma separated.  If multiple search terms are used then objects will be returned in the list only if all the provided terms are matched.
+By default, searches will use case-insensitive partial matches. The search parameter may contain multiple search terms, which should be whitespace and/or comma separated .  If multiple search terms are used then objects will be returned in the list only if all the provided terms are matched. Searches may contain _quoted phrases_ with spaces, each phrase is considered as a single search term.
```
```suggestion
By default, searches will use case-insensitive partial matches. The search parameter may contain multiple search terms, which should be whitespace and/or comma separated. If multiple search terms are used then objects will be returned in the list only if all the provided terms are matched. Searches may contain _quoted phrases_ with spaces, each phrase is considered as a single search term.
```

auvipy commented:
do you intend to add anything else here or improve anything else?

sevdog commented:
As of now I belive that this should be fine.

I tried to figure out what could have been added to the docs, but I felt that it was just copy-pasting fron django admin docs without adding any value.

tomchristie commented:
Really nice bit of functionality, thanks... have followed up with #9338.
